### ISSUE TITLE
Include tqdm in requirements?

### ISSUE THREAD

**Issue by dlakaplan** [nanograv/PINT#1144]
`tqdm` (https://tqdm.github.io) is a nice progress-bar package, but is not currently part of PINT.  I think it's what `emcee` uses for its progress bars (https://github.com/dfm/emcee/blob/main/src/emcee/pbar.py) which we use but don't require.  I was looking to use it as a way to show computation progress for gridding calculations.  I will make it optional for now but we might want to consider making it a requirement (or a similar package).

**Comment by aarchiba** [comment]
Astropy has a similar setup for progress bars, notably for parallelized downloading of cache material. I think it's `tqdm` but it's worth checking - using the same one as astropy would be useful.

**Comment by dlakaplan** [comment]
Looks like astropy does their own:
https://github.com/astropy/astropy/blob/main/astropy/utils/console.py
I don't know how it's better/different/worse.  I used tqdm since that's what emcee used, but I can make astropy the fallback or use only that version.

**Comment by aarchiba** [comment]
Since astropy is already a dependency maybe let's use that? I seem to recall them discussing a switch to tqdm as well...

**Comment by aarchiba** [comment]
https://docs.astropy.org/en/stable/api/astropy.utils.console.ProgressBar.html#astropy.utils.console.ProgressBar Theirs seems to include using multiprocessing to carry out a map, with progress reporting. 

There's a minor consideration with parallel maps - the jobs generically finish in arbitrary order, so you might have finished forty jobs but you're still waiting for job #2 to finish; this is fine, but the progress bar will look weird if it counts only jobs in order. There are "unordered map" interfaces to work around this but it's not clear that that's worth it for progress bars.

**Comment by dlakaplan** [comment]
Yes, and tqdm had a similar interface.  But I think the astropy one just uses the `multiprocessing` module (that we've moved away from).  Using their built-in map methods may make it harder for us to use our own executors

I'm also less certain about the astropy interface, but I'll keep on it.  It might require some significant re-writes.

**Comment by aarchiba** [comment]
For what it's worth, I think that perhaps the only place astropy itself uses the parallel stuff might be in the multiple-downloads interface, which doesn't need any kind of multiprocessing because it's just parallelizing waits for the network.

**Comment by paulray** [comment]
I just added this to polycos.py so that it will use tqdm if it is there and just have no progress bar if it is not. Is this fine?
```
try:
    from tqdm import tqdm
except (ModuleNotFoundError, ImportError) as e:

    def tqdm(*args, **kwargs):
        if args:
            return args[0]
        return kwargs.get("iterable", None)
```

**Comment by dlakaplan** [comment]
yes, I think so

**Comment by abhisrkckl** [comment]
Can this be closed?
